build.sbt

```diff
@@ -184,18 +203,61 @@ val `polynote-server` = project.settings(
 ).dependsOn(`polynote-runtime` % "provided", `polynote-runtime` % "test", `polynote-kernel` % "provided", `polynote-kernel` % "test->test")
 
 val sparkSettings = Seq(
+  resolvers ++= {
+    Seq(MavenRepository(name = "Apache Snapshots", root = "https://repository.apache.org/content/repositories/snapshots"))
+  },
   sparkVersion := {
-    scalaVersion.value match {
-      case ver if ver startsWith "2.11" => "2.1.1"
-      case ver                          => "2.4.7"  // Spark 2.4 is first version to publish for scala 2.12
+    scalaBinaryVersion.value match {
+      case "2.11" => "2.1.1"
+      case "2.12" => "3.1.2"
+      case "2.13" => "3.2.0-SNAPSHOT" // No stable release yet, 3.2.0-SNAPSHOT is earliest available for 2.13
+    }
+  },
+  sparkDistVersion := {
+    sparkVersion.value match {
+      case two if two startsWith "2" => "2.1.1"
+      case three                     => "3.1.2"
     }
   },
   libraryDependencies ++= Seq(
     "org.apache.spark" %% "spark-sql" % sparkVersion.value % "provided",
     "org.apache.spark" %% "spark-repl" % sparkVersion.value % "provided",
     "org.apache.spark" %% "spark-sql" % sparkVersion.value % "test",
     "org.apache.spark" %% "spark-repl" % sparkVersion.value % "test"
-  )
+  ),
+  sparkInstallLocation := {
+    sys.env.get("SPARK_INSTALL_LOCATION")
+      //.orElse(sys.env.get("SPARK_HOME").map(file).map(_.getParent))
+      .getOrElse((file(".").getAbsoluteFile / "target" / "spark").toString)
+  },
+  Test / testOptions += Tests.Setup { () =>
```

Comment:
Curious, why did you move the Spark setup stuff out of the Dockerfile?

Also, will this setup happen if tests aren't run?